File: app/src/main/java/com/smile/watchmovie/fragment/SearchFragment.java
package com.smile.watchmovie.fragment;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.SearchView;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.smile.watchmovie.adapter.FilmSearchAdapter;
import com.smile.watchmovie.custom.PaginationScrollListener;
import com.smile.watchmovie.api.ApiService;
import com.smile.watchmovie.databinding.FragmentSearchBinding;
import com.smile.watchmovie.model.FilmArrayResponse;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class SearchFragment extends Fragment {
    private FragmentSearchBinding binding;
    private FilmSearchAdapter mFilmSearchAdapter;
    private boolean mIsLoading, mIsLastPage;
    private int currentPage = 0;
    private String keySearch;

    @SuppressLint("NotifyDataSetChanged")
    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentSearchBinding.inflate(inflater, container, false);

        mFilmSearchAdapter = new FilmSearchAdapter(requireActivity());
        binding.rcvFilm.setAdapter(mFilmSearchAdapter);
        binding.searchView.clearFocus();

        binding.searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @SuppressLint("SetTextI18n")
            @Override
            public boolean onQueryTextSubmit(String query) {
                if (!query.trim().equals("")) {
                    keySearch = query;
                    currentPage = 0;
                    binding.tvTitle.setVisibility(View.GONE);
                    mFilmSearchAdapter.clearData();
                    binding.loadSearchPage.setVisibility(View.VISIBLE);
                    callApiSearchFilm(query, 0);
                }
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                return false;
            }
        });

        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(requireActivity(), RecyclerView.VERTICAL, false);
        binding.rcvFilm.setLayoutManager(linearLayoutManager);
        binding.rcvFilm.addOnScrollListener(new PaginationScrollListener(linearLayoutManager) {
            @Override
            public void loadMoreItems() {
                mIsLoading = true;
                currentPage += 1;
                loadNextPage();
            }

            @Override
            public boolean isLoading() {
                return mIsLoading;
            }

            @Override
            public boolean isLastPage() {
                return mIsLastPage;
            }
        });

        return binding.getRoot();
    }

    private void callApiSearchFilm(String key, int page) {
        ApiService.apiService.searchFilms("7da353b8a3246f851e0ee436d898a26d", key, page, 10).enqueue(new Callback<FilmArrayResponse>() {
            @Override
            public void onResponse(@NonNull Call<FilmArrayResponse> call, @NonNull Response<FilmArrayResponse> response) {
                binding.loadSearchPage.setVisibility(View.GONE);
                if (response.body() != null) {
                    mFilmSearchAdapter.updateData(response.body().getData());
                }
            }

            @Override
            public void onFailure(@NonNull Call<FilmArrayResponse> call, @NonNull Throwable t) {
                binding.loadSearchPage.setVisibility(View.GONE);
                binding.tvTitle.setVisibility(View.VISIBLE);
                Toast.makeText(requireActivity(), "Tìm kiếm không thành công", Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void loadNextPage() {
        mIsLoading = false;
        callApiSearchFilm(keySearch, currentPage);
        if (currentPage == 20) {
            mIsLastPage = true;
        }
    }
}

File: app/src/main/java/com/smile/watchmovie/fragment/SearchFragment.kt
package com.smile.watchmovie.fragment

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.widget.SearchView
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.paging.LoadState
import com.smile.watchmovie.activity.PlayerActivity
import com.smile.watchmovie.adapter.PagingAdapter
import com.smile.watchmovie.api.ApiHelper
import com.smile.watchmovie.api.RetrofitBuilder
import com.smile.watchmovie.base.ViewModelFactory
import com.smile.watchmovie.databinding.FragmentSearchBinding
import com.smile.watchmovie.model.FilmMainHome
import com.smile.watchmovie.utils.Constant
import com.smile.watchmovie.viewmodel.SearchFragmentViewModel
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

class SearchFragment : Fragment(), PagingAdapter.OnClickListener {
    private lateinit var binding: FragmentSearchBinding
    private lateinit var viewModel: SearchFragmentViewModel
    private lateinit var pagingAdapter: PagingAdapter

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentSearchBinding.inflate(inflater, container, false)

        initializeData()
        handleEvent()

        return binding.root
    }

    private fun initializeData() {
        viewModel = ViewModelProvider(
            this,
            ViewModelFactory(ApiHelper(RetrofitBuilder.apiServiceFilm))
        )[SearchFragmentViewModel::class.java]

        pagingAdapter = PagingAdapter(requireActivity(), this)

        binding.apply {
            searchView.clearFocus()

            rcvFilm.apply {
                adapter = pagingAdapter
                setHasFixedSize(true)
            }
        }
    }

    private fun handleEvent() {
        binding.searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                if (query.trim().isNotEmpty()) {
                    viewLifecycleOwner.lifecycleScope.launch {
                        viewModel.getListFilm(query).collectLatest {
                            pagingAdapter.submitData(it)
                        }
                    }
                }
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {
                return false
            }
        })

        pagingAdapter.addLoadStateListener {
            if (it.refresh == LoadState.Loading) {
                binding.apply {
                    progressBar.visibility = View.VISIBLE
                    tvTitle.visibility = View.GONE
                }
            } else if (it.prepend.endOfPaginationReached && pagingAdapter.itemCount < 1) {
                binding.apply {
                    progressBar.visibility = View.GONE
                    tvTitle.visibility = View.VISIBLE
                }
            } else {
                binding.progressBar.visibility = View.GONE
            }
        }
    }

    override fun playFilm(film: FilmMainHome) {
        viewModel.getFilmDetail(Constant.Api.WS_TOKEN, film.id).observe(requireActivity()) {
            it?.let {
                val intent = Intent(requireActivity(), PlayerActivity::class.java)
                intent.putExtra("film", it.data)
                startActivity(intent)
            }
        }
    }
}